signac/core/jsoncollection.py

```diff
@@ -0,0 +1,142 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+import os
+import json
+import errno
+import uuid
+
+from .collection_api import SyncedCollection
+from .syncedattrdict import SyncedAttrDict
+from .syncedlist import SyncedList
+
+
+class JSONCollection(SyncedCollection):
+    """Implement sync and load using a JSON back end."""
+
+    backend = 'JSON'  # type: ignore
+
+    def __init__(self, filename=None, write_concern=False, **kwargs):
+        self._filename = os.path.realpath(filename) if filename is not None else None
+        self._write_concern = write_concern
+        super().__init__(**kwargs)
+        if (filename is None) == (self._parent is None):
+            raise ValueError(
+                "Illegal argument combination, one of the two arguments, "
+                "parent or filename must be None, but not both.")
+
+    def _load(self):
+        "Loads the data from json file"
+        try:
+            with open(self._filename, 'rb') as file:
+                blob = file.read()
+                return json.loads(blob.decode())
+        except IOError as error:
+            if error.errno == errno.ENOENT:
+                return None
+
+    def _sync(self):
+        "Write the data to json file"
+        data = self.to_base()
+        # Serialize data:
+        blob = json.dumps(data).encode()
+
+        if self._write_concern:
+            dirname, filename = os.path.split(self._filename)
+            fn_tmp = os.path.join(dirname, '._{uid}_{fn}'.format(
+                uid=uuid.uuid4(), fn=filename))
+            with open(fn_tmp, 'wb') as tmpfile:
+                tmpfile.write(blob)
+            os.replace(fn_tmp, self._filename)
+        else:
+            with open(self._filename, 'wb') as file:
+                file.write(blob)
+
+
+class JSONDict(JSONCollection, SyncedAttrDict):
+    """A dict-like mapping interface to a persistent JSON file.
+
+    The JSONDict is a :class:`~collections.abc.MutableMapping` and therefore
+    behaves similar to a :class:`dict`, but all data is stored persistently in
+    the associated JSON file on disk.
```

Comment:
We should very clearly split the docstrings between the different classes. Specifically, the description of persistent file storage at a high level should be at the SyncedCollection level (I think that class needs to have an extensive docstring). Every specific collection type (e.g. JSONCollection) should document the backend it implements, and every data type (e.g. SyncedDict) should document the core Python data type it mimics. Then the bottom level classes (e.g. JSONDict) will just have very short descriptions and point to their parents for documentation on *how* this is implemented (since that's where the implementation will be).

Reply:
Having said that, the examples below are good to have here, and I'm fine with duplicating warnings if we think that's necessary (although we can decide that later).